src/vumeter-qt/vumeter.cc

```diff
@@ -0,0 +1,437 @@
+}
+
+void VUMeterQtWidget::draw_visualizer_peaks(QPainter &p)
+{
+    float bar_width = get_bar_width(nchannels);
+    float font_size_width = bar_width / 3.0f;
+    float font_size_height = vumeter_top_padding * 0.50f;
+
+    QFont font = p.font();
+    font.setPointSizeF(fminf(font_size_width, font_size_height));
+    p.setFont(font);
+
+    QPen pen = p.pen();
+    pen.setColor(text_color);
+    p.setPen(pen);
+
+    QFontMetricsF fm(p.font());
+    char text[10];
+    for (int i = 0; i < nchannels; i++)
+    {
+        format_db(text, channels_peaks[i]);
+        QSizeF text_size = fm.size(0, text);
+        p.drawText(QPointF(legend_width + bar_width*(i+0.5f) - text_size.width()/2.0f,
+            vumeter_top_padding/2.0f + (text_size.height()/4.0f)), text
+        );
+    }
+}
+
+void VUMeterQtWidget::draw_visualizer(QPainter & p)
+{
+    for (int i = 0; i < nchannels; i++)
+    {
+        float bar_width = get_bar_width(nchannels);
+        float x = legend_width + (bar_width * i);
+        if (i > 0)
+        {
+             x += 1;
+             bar_width -= 1;
+        }
+
+        p.fillRect (
+            QRectF(x, vumeter_top_padding, bar_width, vumeter_height),
+            background_vumeter_pattern
+        );
+
+        p.fillRect (
+            QRectF(x, get_y_from_db(channels_db_level[i]),
+                bar_width, (get_height_from_db(channels_db_level[i]))),
+            vumeter_pattern
+        );
+
+        if (channels_peaks[i] > -VUMeterQt::db_range)
+        {
+            p.fillRect (
+                QRectF(x, get_y_from_db(channels_peaks[i]), bar_width, 1),
+                vumeter_pattern
+            );
+        }
+    }
+}
+
+void VUMeterQtWidget::format_db(char *buf, const float val) {
+    if (val > -10)
+    {
+        sprintf(buf, "%+.1f", val);
+    }
+    else if (val > -VUMeterQt::db_range)
+    {
+        sprintf(buf, "%.0f ", val);
+    }
+    else
+    {
+        sprintf(buf, "-inf");
+    }
+}
+
+float VUMeterQtWidget::get_bar_width(int channels)
+{
+    return vumeter_width / channels;
+}
+
+void VUMeterQtWidget::update_sizes()
+{
+    if (height() > 200 && width() > 60 && aud_get_bool("vumeter", "display_legend")) {
+        must_draw_vu_legend = true;
+        vumeter_top_padding = height() * 0.03f;
+        vumeter_bottom_padding = height() * 0.015f;
+        vumeter_height = height() - vumeter_top_padding - vumeter_bottom_padding;
+        legend_width = width() * 0.3f;
+        vumeter_width = width() - (legend_width * 2);
+    } else {
+        must_draw_vu_legend = false;
+        vumeter_top_padding = 0;
+        vumeter_bottom_padding = 0;
+        vumeter_height = height();
+        legend_width = 0;
+        vumeter_width = width();
+    }
+    vumeter_pattern = get_vumeter_pattern();
+    background_vumeter_pattern = get_vumeter_pattern(30);
+}
+
+VUMeterQtWidget::VUMeterQtWidget (QWidget * parent) : QWidget (parent)
+{
+    update_sizes();
+
+    setObjectName ("VUMeterQtWidget");
```

Comment:
this shouldn't be necessary, it will automatically infer this object name from the class name